crates/bevy_render/src/render_resource/storage_buffer.rs

```diff
@@ -1,141 +1,153 @@
+use encase::{
+    internal::WriteInto, DynamicStorageBuffer as DynamicStorageBufferWrapper, ShaderType,
+    StorageBuffer as StorageBufferWrapper,
+};
+use wgpu::{util::BufferInitDescriptor, BindingResource, BufferBinding, BufferUsages};
+
+pub struct StorageBuffer<T: ShaderType> {
+    value: T,
+    scratch: StorageBufferWrapper<Vec<u8>>,
+    buffer: Option<Buffer>,
+    capacity: usize,
 }
 
-impl<T: AsStd430, U: AsStd430 + Default> Default for StorageBuffer<T, U> {
-    /// Creates a new [`StorageBuffer`]
-    ///
-    /// This does not immediately allocate system/video RAM buffers.
-    fn default() -> Self {
+impl<T: ShaderType> From<T> for StorageBuffer<T> {
+    fn from(value: T) -> Self {
         Self {
-            body: U::default(),
-            values: Vec::new(),
-            scratch: Vec::new(),
-            storage_buffer: None,
+            value,
+            scratch: StorageBufferWrapper::new(Vec::new()),
+            buffer: None,
+            capacity: 0,
         }
     }
 }
 
-impl<T: AsStd430, U: AsStd430> StorageBuffer<T, U> {
-    // NOTE: AsStd430::std430_size_static() uses size_of internally but trait functions cannot be
-    // marked as const functions
-    const BODY_SIZE: usize = std::mem::size_of::<U>();
-    const ITEM_SIZE: usize = std::mem::size_of::<T>();
+impl<T: ShaderType + Default> Default for StorageBuffer<T> {
+    fn default() -> Self {
+        Self {
+            value: T::default(),
+            scratch: StorageBufferWrapper::new(Vec::new()),
+            buffer: None,
+            capacity: 0,
+        }
+    }
+}
 
-    /// Gets the reference to the underlying buffer, if one has been allocated.
+impl<T: ShaderType + WriteInto> StorageBuffer<T> {
     #[inline]
     pub fn buffer(&self) -> Option<&Buffer> {
-        self.storage_buffer.as_ref()
+        self.buffer.as_ref()
     }
 
     #[inline]
     pub fn binding(&self) -> Option<BindingResource> {
-        Some(BindingResource::Buffer(BufferBinding {
-            buffer: self.buffer()?,
-            offset: 0,
-            size: Some(NonZeroU64::new((self.size()) as u64).unwrap()),
-        }))
+        Some(BindingResource::Buffer(
+            self.buffer()?.as_entire_buffer_binding(),
+        ))
     }
 
-    #[inline]
-    pub fn set_body(&mut self, body: U) {
-        self.body = body;
+    pub fn set(&mut self, value: T) {
+        self.value = value;
     }
 
-    fn reserve_buffer(&mut self, device: &RenderDevice) -> bool {
-        let size = self.size();
-        if self.storage_buffer.is_none() || size > self.scratch.len() {
-            self.scratch.resize(size, 0);
-            self.storage_buffer = Some(device.create_buffer(&BufferDescriptor {
+    pub fn get(&self) -> &T {
+        &self.value
+    }
+
+    pub fn get_mut(&mut self) -> &mut T {
+        &mut self.value
+    }
+
+    pub fn write_buffer(&mut self, device: &RenderDevice, queue: &RenderQueue) {
+        self.scratch.write(&self.value).unwrap();
```

Comment:
How does this handle if there is a `Vec` in the middle of the struct, as in, not a trailing variable-sized array? Does it just assume that its number of elements in the `Vec` matches the size of the array in the shader?

Reply:
That won't compile. The proc macro will error since the `#[size(runtime)]` attribute can only be used on the last field of a struct.

Reply:
Excellent!